src/Symfony/Component/DependencyInjection/Attribute/When.php

```diff
@@ -0,0 +1,26 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Attribute;
+
+/**
+ * An attribute to tell under which environement this class should be autoconfigured.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+#[\Attribute(\Attribute::TARGET_CLASS | \Attribute::IS_REPEATABLE)]
+class When
```

Comment:
If this is all about autoconfiguration, then should it be `AutoconfigureWhen(env: 'foo')`? Just `When` makes it seem like perhaps you are able to conditionally register the service itself.

Reply:
> `AutoconfigureWhen`

We did not add `Auto` to eg `AsCommand`, `AsEventListener`, etc, while they apply only when autoconfiguration happens. I now think we did right, and that the `Auto-anything` prefix would be redundant with the very nature of attributes, which is that they are declarative and that they always need an "auto-something" to be read. If something is always needed, we can omit, it since it doesn't bring any useful info on its own.

In this specific case, there is something more: the attribute is applied only when autoconfiguring *and* autodiscovering.

The reason is that when one writes:

```yaml
services:
    App\Foo:
        autoconfigure: true # this being either explicit, or coming from _defaults
```

Then one won't expect the `When` attribute to be taken into account.

So to me, the attribute should apply only when doing autodiscovery:
yaml
```
services:
    App\:
        resources: ../src
```

Reply:
> Just When makes it seem like perhaps you are able to conditionally register the service itself.

Because services are private, this will end up being the case. E.g. if a command is not tagged as `console.command`, it's going to be removed from the container.

But we could also skip registering the service entirely. There are subtle differences between the two approaches. I'm not sure which one has the least surprise effect. Your comment makes me think it could be better to just skip registering the corresponding classes.

I'm going to update the PR.

Reply:
PR updated: classes will now simply be not registered during autodiscovery if the attribute doesn't match the current env.